src/types/typeGuards/compound.ts

```diff
@@ -1,7 +1,104 @@
 import * as ts from "typescript";
 
+import { isTypeFlagSet } from "../../flags.js";
 import { isTupleType, isTypeReference } from "./objects.js";
 
+/**
+ * `LiteralType` from typescript except that it allows for it to work on arbitrary types.
+ *
+ * @category Type Types
+ */
+export interface GeneralLiteralType extends ts.Type {
+	freshType: GeneralLiteralType;
+	regularType: GeneralLiteralType;
+	value: unknown;
+}
+
+/**
+ * A boolean literal.
+ *
+ * @category Type Types
+ */
+export interface BooleanLiteralType extends GeneralLiteralType {
+	value: boolean;
+}
+
+/**
+ * A "true" literal.
+ *
+ * @category Type Types
+ */
+
+export interface TrueLiteralType extends BooleanLiteralType {
+	value: true;
+}
+
+/**
+ * A "false" literal.
+ *
+ * @category Type Types
+ */
+export interface FalseLiteralType extends BooleanLiteralType {
+	value: false;
+}
+
+/**
+ * Determines whether the given type is a boolean literal type.
+ *
+ * @category Types - Utilities
+ */
+export function isBooleanLiteralType(type: ts.Type): type is BooleanLiteralType;
+
+/**
+ * Determines whether the given type is a boolean literal type for "true".
+ *
+ * @category Types - Utilities
+ */
+export function isBooleanLiteralType(
+	type: ts.Type,
+	literal: true
+): type is TrueLiteralType;
+
+/**
+ * Determines whether the given type is a boolean literal type for "false".
+ *
+ * @category Types - Utilities
+ */
+export function isBooleanLiteralType(
+	type: ts.Type,
+	literal: false
+): type is FalseLiteralType;
+
+export function isBooleanLiteralType(
+	type: ts.Type,
+	literal?: boolean
+): boolean {
+	return (
+		isTypeFlagSet(type, ts.TypeFlags.BooleanLiteral) &&
+		(literal === undefined ||
+			(type as unknown as { intrinsicName: string }).intrinsicName ===
+				(literal ? "true" : "false"))
+	);
+}
+
+/**
+ * Determines whether the given type is a boolean literal type for "true".
+ *
+ * @category Types - Utilities
+ */
+export function isTrueLiteralType(type: ts.Type): type is TrueLiteralType {
```

Comment:
Do we want this function (and the one below) or do we just want the overloads of `isBooleanLiteralType`.

Reply:
I see no strong reason to keep the extra functions when `isBooleanLiteralType` works for the same purposes. Let's delete. 🔪

Reply:
We could alternatively remove the `literal` parameter from `isBooleanLiteralType` (and the overloads).

Reply:
🤷 no strong preference from me, up to you!